### DISCUSSION TITLE
UI Mockup: End Time feature for Temporal Drawing Tools

### DISCUSSION THREAD

**Discussion by ErikOstlund** [zooniverse/front-end-monorepo#2270] [Ideas]
A requested feature for our Temporal Drawing Tools is to add the ability for users to add an End Time.

As a reminder, when a Temporal Drawing Tool is created on a video, the video timestamp is the Display Time. The annotation will not display on the video until that moment in the video.

The End Time will be used to represent the moment in a video where any point of interest is no longer visible. For example, a Solar jet ends and is no longer visible in the video.
Creating an EndTime is optional.
When an End Time is created, the annotation will display on the video from DisplayTime to EndTime.

This is a possible UI solution to make this feature user friendly and intuitive.
All feedback is welcome! 

These UI mockups were created using fluidui.com. Not a bad mockup/wireframe app; all online, nothing to install. But the free edition only allows 1 project.

![image](https://user-images.githubusercontent.com/9533007/123457249-55c38b80-d5a9-11eb-8d0c-f85fc16bc921.png)

![image](https://user-images.githubusercontent.com/9533007/123457328-6ecc3c80-d5a9-11eb-86d9-da2c567ba459.png)

![image](https://user-images.githubusercontent.com/9533007/123457381-7b509500-d5a9-11eb-9355-8f015b3e713b.png)

![image](https://user-images.githubusercontent.com/9533007/123457438-886d8400-d5a9-11eb-9243-d9ea470cd9a8.png)

![image](https://user-images.githubusercontent.com/9533007/123457487-96230980-d5a9-11eb-8c4d-f42d6773281d.png)

**Comment by shaunanoordin** [comment]
Quick thoughts (in no particular order)

- Main point: **LGTM!** 👍 I like the idea as presented in general, and everything else in the following list will be detailed nitpicking.
- That list of annotations (on the right-hand Task Area) has the possibility of going reaaallly long...
  - ...but that depends on the project. For Solar Jets (and for foreseeable future projects) this shouldn't be an issue.
  - 👻 ...but if we DO want to plan for potential future scenarios, one way to handle this is to wrap the list of annotations (right-hand Task Area) in a scrollable Div with a max height, or similar. This will have its own issues though, like trying to highlight which annotation currently has focus. (See below)
- Once you have more than 1-2 marks per Temporal Drawing Mark type, it's going to be very confusing to understand which mark (on the left-hand Subject Viewer) corresponds to which annotation (on the right-hand Task Area)
  - e.g. There are 2x Temporal Points in the examples above. Which one corresponds to 01:250?
  - 💡 UI improvement idea: I should be able to click on a mark (left-hand Subject Viewer) to make it the **Active Mark,** and **the corresponding annotation (right-hand Task Area) should be highlighted.**
  - 💡 Conversely: clicking on an annotation (right-hand Task Area) should **highlight it,** and make its corresponding mark (left-hand Subject viewer) the **Active Mark.**
- UI language: The trashcan icon next to each annotation (Task Area) should probably be **the same ✖️ icon** used to delete marks (Subject Viewer)
- 👻 If a user can ADD an End Time, should they also be allowed to DELETE or MODIFY the End Time without having to delete the whole mark/annotation?
  - 👻 If the answer is _yes,_ then adding a new "delete end time" or "modify end time" buttons to the Task Area without making the UI too cluttered or confusing is going to be _a confounding challenge, at best._
  - 👻 If the answer is yes _and_ you somehow figure out how to add "delete End Time" or "Modify End Time" buttons, you know that somebody's going to ask for a "Modify START TIME" button as well.
  - 👻 👻 👻 
- 👻 If this UI idea takes off, I can foresee project owners asking "hey, can we get that List of Annotations" for non-Temporal drawing tools too??" ...which is going to be its own can of worms!

Anything marked with 👻  indicates "BEWARE OF FEATURE CREEP". Take those ideas/thoughts/suggestions with a pinch of salt and be prepared to just respond with _"actually let's not worry about this item right now",_ as I may be worrying about larger-view issues that may be beyond the intended scope of your work or this discussion.

**Comment by goplayoutside3** [comment]
I like this general layout, and also agree with the points Shaun made! 

My one suggestion is to consistently label 'Display Time' and 'End Time' in the UI. For example, 'End Time' should not change to 'Hide Time'. I think a user making classifications on a project with this tool will think in basic terms of "...this is when the data **_starts_** and **_ends_**" as opposed to "...this is when the data **_displays_** and **_hides_**". I suggest labeling time measurements as 'Start Time' and 'End Time' in the UI.

**Comment by srallen** [comment]
I think this is potential idea, but I think there are some pros and cons to consider:

- This may be a good way to indicate these a temporal drawing tools. I still think we need some kind of visual indicator to distinguish between the standard tools.
- I agree with @shaunanoordin that if the list of marks in the task area allows you to edit the end time, then users will probably expect that you can edit start time.
- I agree with @goplayoutside3 that the language should be 'start' and 'end'

However, I'm wondering if we could consider an alternative that users might be familiar with in our UI. Being able to edit the start and end time could be wired up to be a "sub-task" even though technically these annotations aren't. We can reuse the UI and functionality of the sub-task pop-up for each mark to render two inputs for start and end time. When the mark is made the start time is already filled with the time the video is paused and the end time auto-filled in with the end of the video time. The user will can edit and return to edit (when the mark is visible) by clicking on the mark much like they do when creating and editing sub-tasks. 

A potential problem with either of these UIs is that if the user edits the start or end time such that it should no longer display at the time the video is paused. If they do, then it might be confusing why the mark disappears?
